cpp/src/datetime/datetime_ops.cu

```diff
@@ -89,9 +85,28 @@ struct extract_component_operator {
         return static_cast<unsigned>(year_month_day(days_since_epoch).day());
       case datetime_component::WEEKDAY:
         return year_month_weekday(days_since_epoch).weekday().iso_encoding();
-      case datetime_component::HOUR: return hrs_.count();
-      case datetime_component::MINUTE: return mins_.count();
-      case datetime_component::SECOND: return secs_.count();
+      case datetime_component::HOUR:
+      case datetime_component::MINUTE:
+      case datetime_component::SECOND:
+      case datetime_component::MILLISECOND:
+      case datetime_component::MICROSECOND:
+      case datetime_component::NANOSECOND: {
+        auto const hrs_ = duration_cast<hours>(time_since_midnight);
+        if (Component == datetime_component::HOUR) { return hrs_.count(); }
+        auto const mins_ = duration_cast<minutes>(time_since_midnight - hrs_);
+        if (Component == datetime_component::MINUTE) { return mins_.count(); }
+        auto const secs_ = duration_cast<seconds>(time_since_midnight - hrs_ - mins_);
+        if (Component == datetime_component::SECOND) { return secs_.count(); }
+        auto const millisecs_ =
+          duration_cast<milliseconds>(time_since_midnight - hrs_ - mins_ - secs_);
+        if (Component == datetime_component::MILLISECOND) { return millisecs_.count(); }
+        auto const microsecs_ =
+          duration_cast<microseconds>(time_since_midnight - hrs_ - mins_ - secs_ - millisecs_);
+        if (Component == datetime_component::MICROSECOND) { return microsecs_.count(); }
+        auto const nanosecs_ = duration_cast<nanoseconds>(time_since_midnight - hrs_ - mins_ -
+                                                          secs_ - millisecs_ - microsecs_);
+        if (Component == datetime_component::NANOSECOND) { return nanosecs_.count(); }
```

Comment:
Not a sticking point for me, but now we've gone to checking `component == HOUR` twice: once for the switch case, and then again for the explicit `if`. Any advantage of a switch-table is lost.
I wonder if there was value in simply converting the previous definitions to lambdas:
```c++
auto const hrs_  = [&] { return duration_cast<hours>(time_since_midnight); };
auto const mins_ = [&] { return duration_cast<minutes>(time_since_midnight) - hrs_(); };
// ...
```
The switch-case would now look like:
```c++
switch(component)
{
  // ...
  case datetime_component::HOUR: return hrs_().count();
  case datetime_component::MINUTE: return mins_().count();
  // et cetera.
}
```

Reply:
@galipremsagar, please let me know if this looks agreeable. It should address @ttnghia's concern, without the bloat, or the unnecessary computation.

Reply:
Seems reasonable to me, made the change in https://github.com/rapidsai/cudf/pull/11911/commits/9c8dbfb16951a27df9a0d8e666108c429ebc4be1